packages/@aws-cdk/cloudformation-diff/lib/diff/template-and-changeset-diff-merger.ts

```diff
@@ -0,0 +1,139 @@
+        propertyReplacementModes: propertyReplacementModes,
+      };
+    }
+
+    return changeSetResources;
+  }
+
+  public overrideDiffResourceChangeImpactWithChangeSetChangeImpact(logicalId: string, change: types.ResourceDifference) {
+    // resourceType getter throws an error if resourceTypeChanged
+    if ((change.resourceTypeChanged === true) || change.resourceType?.includes('AWS::Serverless')) {
+      // CFN applies the SAM transform before creating the changeset, so the changeset contains no information about SAM resources
+      return;
+    }
+    change.forEachDifference((type: 'Property' | 'Other', name: string, value: types.Difference<any> | types.PropertyDifference<any>) => {
+      if (type === 'Property') {
+        if (!this.changeSetResources[logicalId]) {
+          (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.NO_CHANGE;
+          (value as types.PropertyDifference<any>).isDifferent = false;
+          return;
+        }
+
+        const changingPropertyCausesResourceReplacement = (this.changeSetResources[logicalId].propertyReplacementModes ?? {})[name]?.replacementMode;
+        switch (changingPropertyCausesResourceReplacement) {
+          case 'Always':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.WILL_REPLACE;
+            break;
+          case 'Never':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.WILL_UPDATE;
+            break;
+          case 'Conditionally':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.MAY_REPLACE;
+            break;
+          case undefined:
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.NO_CHANGE;
+            (value as types.PropertyDifference<any>).isDifferent = false;
+            break;
+          // otherwise, defer to the changeImpact from the template diff
+        }
+      } else if (type === 'Other') {
+        switch (name) {
+          case 'Metadata':
+            // we want to ignore metadata changes in the diff, so compare newValue against newValue.
+            change.setOtherChange('Metadata', new types.Difference<string>(value.newValue, value.newValue));
+            break;
+        }
+      }
+    });
+  }
+
+  public addImportInformationFromChangeset(resourceDiffs: types.DifferenceCollection<types.Resource, types.ResourceDifference>) {
+    const imports = this.findResourceImports();
+    resourceDiffs.forEachDifference((logicalId: string, change: types.ResourceDifference) => {
+      if (imports.includes(logicalId)) {
+        change.isImport = true;
+      }
+    });
+  }
+
+  public findResourceImports(): string[] {
+    const importedResourceLogicalIds = [];
+    for (const resourceChange of this.changeSet?.Changes ?? []) {
```

Comment:
This is just me being extra defensive... we could get rid of that, but I'd rather just be overly cautious?

Reply:
I don't think its needed though since you can't construct the class without a `changeSet` provided. Because of that we know `changeSet` can never be undefined.